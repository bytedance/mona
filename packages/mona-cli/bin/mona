#! /usr/bin/env node

const mona = require('../dist').default;
mona();
