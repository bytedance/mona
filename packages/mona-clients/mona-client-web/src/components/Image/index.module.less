.wrapper {
  width: 300px;
  height: 225px;
  display: inline-block;
  overflow: hidden;

  .hidden {
    width: 0;
    height: 0;
    position: absolute;
    opacity: 0;
  }

  .image {
    width: 100%;
    height: 100%;
    background-repeat: no-repeat;
    background-size: 100% 100%;
    background-position: center center;

    &.scaleToFill {
      background-size: 100% 100%;
    }

    &.aspectFit {
      background-size: contain;
    }

    &.aspectFill {
      background-size: cover;
    }

    // TODO
    &.widthFix {
      background-size: auto;
    }

    &.heightFix {
      background-size: auto;
    }

    &.top {
      background-size: auto;
      background-position: center top;
    }

    &.bottom {
      background-size: auto;
      background-position: center bottom;
    }

    &.center {
      background-size: auto;
      background-position: center center;
    }

    &.left {
      background-size: auto;
      background-position: left center;
    }

    &.right {
      background-size: auto;
      background-position: right center;
    }

    &.topleft {
      background-size: auto;
      background-position: left top;
    }

    &.topright{
      background-size: auto;
      background-position: right top;
    }

    &.bottomleft{
      background-size: auto;
      background-position: left bottom;
    }

    &.bottomright {
      background-size: auto;
      background-position: right bottom;
    }
  }
}