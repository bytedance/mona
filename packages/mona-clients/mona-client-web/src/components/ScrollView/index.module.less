.container {
  position: relative;
  -webkit-overflow-scrolling: touch;
  height: 100%;
  width: 100%;
  max-height: inherit;

  &::-webkit-scrollbar {
    display: none;
  }
}

