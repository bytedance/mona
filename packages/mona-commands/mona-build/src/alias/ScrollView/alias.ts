import { bubbleEventsAlias, basePropAlias, baseComponentPropsMap as aliasMap } from '../prop';
import { ScrollViewAlias as Alias } from '../type';

const alias: Alias = {
  ...basePropAlias,
  ...bubbleEventsAlias,
  scrollX: aliasMap.scrollX,
  scrollY: aliasMap.scrollY,
  scrollIntoView: aliasMap.scrollIntoView,
  scrollWithAnimation: aliasMap.scrollWithAnimation,
  scrollLeft: aliasMap.scrollLeft,
  scrollTop: aliasMap.scrollTop,
  upperThreshold: aliasMap.upperThreshold,
  lowerThreshold: aliasMap.lowerThreshold,
  onScroll: aliasMap.onScroll,
  onScrollToLower: aliasMap.onScrollToLower,
  onScrollToUpper: aliasMap.onScrollToUpper,
};

export default alias;
