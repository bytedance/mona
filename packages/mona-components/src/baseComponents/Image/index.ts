import React from 'react';
import createComponent from '../../createComponent';
import { BaseProps, EventHandler } from '../base';

type ImageMode = 'scaleToFill' | 'aspectFit' | 'aspectFill' | 'widthFix' | 'heightFix' | 'top' | 'bottom' | 'left' | 'right' | 'top left' | 'top right' | 'bottom left' | 'bottom right'

export interface ImageProps extends BaseProps {
  src?: string;
  mode?: ImageMode;
  lazyLoad?: boolean;
  onError?: EventHandler;
  onLoad?: EventHandler;
}

export const Image: React.ComponentType<ImageProps> = createComponent<ImageProps>('image')

