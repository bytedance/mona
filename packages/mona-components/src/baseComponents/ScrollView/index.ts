import React from 'react';
import createComponent from '../../createComponent';
import { BaseProps, EventHandler } from '../base';

export interface ScrollViewProps extends BaseProps {
  scrollX?: boolean;
  scrollY?: boolean;
  scrollWithAnimation?: boolean;
  upperThreshold?: number;
  lowerThreshold?: number;
  scrollTop?: number;
  scrollLeft?: number;
  scrollIntoView?: string;
  onScroll?: EventHandler;
  onScrollToUpper?: EventHandler;
  onScrollToLower?: EventHandler;
}

export const ScrollView: React.ComponentType<ScrollViewProps> = createComponent<ScrollViewProps>('scroll-view')
