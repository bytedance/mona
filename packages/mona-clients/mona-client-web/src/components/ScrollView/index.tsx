import React, { useLayoutEffect, useMemo, useRef } from 'react';
import { ScrollViewProps } from '@bytedance/mona';
import styles from './index.module.less';
import { useHandlers } from '../hooks';
import { formatSyntheticEvent } from '../utils';

//t = current time
//b = start value
//c = change in value
//d = duration
function easeInOutQuad(t: number, b: number, c: number, d: number) {
  t /= d/2;
	if (t < 1) return c/2*t*t + b;
	t--;
	return -c/2 * (t*(t-2) - 1) + b;
};


function scrollToWithAnimation(element: HTMLDivElement, to: number, duration: number, vertical: boolean) {
    const start = vertical ? element.scrollTop : element.scrollLeft;
    const change = to - start;
    let currentTime = 0;
    const increment = 20;
        
    const animateScroll = function(){        
        currentTime += increment;
        const val = easeInOutQuad(currentTime, start, change, duration);
        if (vertical) {
          element.scrollTop = val;
        } else {
          element.scrollLeft = val;
        }
        if(currentTime < duration) {
            setTimeout(animateScroll, increment);
        }
    };
    animateScroll();
}

type onScrollCallback = (ele: Element) => void;

interface ScrollViewContextProps {
  registerScrollCallback: (callback: onScrollCallback) => void;
  unregisterScrollCallback: (callback: onScrollCallback) => void;
}

export const ScrollViewContext = React.createContext<ScrollViewContextProps | null>(null)

const ScrollView: React.FC<ScrollViewProps> = (props) => {
  const {
    children,
    scrollIntoView,
    upperThreshold,
    lowerThreshold,
    scrollWithAnimation,
    scrollTop,
    scrollLeft,
    scrollX,
    scrollY,
    onScroll,
    onScrollToUpper,
    onScrollToLower,
    ...restProps
  } = props;
  const { handleClassName, ...handleProps } = useHandlers(restProps)
  const scrollRef = useRef<HTMLDivElement | null>(null);
  const topPreRef = useRef(0);
  const leftPreRef = useRef(0);

  const cbRefs = useRef<onScrollCallback[]>([]);
  const context = useMemo(() => ({
    registerScrollCallback: (callback: onScrollCallback) => {
      cbRefs.current.push(callback)
    },
    unregisterScrollCallback: (callback: onScrollCallback) => {
      const index = cbRefs.current.indexOf(callback);
      cbRefs.current.splice(index, 1)
    }
  }), [])

  const overflow = `${scrollX ? 'auto' : 'hidden'} ${scrollY ? 'auto' : 'hidden'}`;

  const handleScroll = (e: React.SyntheticEvent) => {
    if (typeof onScroll === 'function') {
      onScroll(formatSyntheticEvent({ event: e }))
    }

    
    if (scrollRef.current) {
      cbRefs.current.forEach(cb => cb(scrollRef.current!));

      const top = scrollRef.current.scrollTop;
      const left = scrollRef.current.scrollLeft;
      const currentUpperThreshold = upperThreshold || 0;
      const currentLowerThreshold = lowerThreshold || 0;

      if (typeof onScrollToUpper === 'function') {
        if (scrollY && ((currentUpperThreshold > topPreRef.current && currentUpperThreshold <= top) || (currentUpperThreshold < topPreRef.current && currentUpperThreshold >= top))) {
          onScrollToUpper(formatSyntheticEvent({ event: e, type: 'scrolltoupper' }))
        }
        if (scrollX && ((currentUpperThreshold > leftPreRef.current && currentUpperThreshold <= left) || (currentUpperThreshold < leftPreRef.current && currentUpperThreshold >= left))) {
          onScrollToUpper(formatSyntheticEvent({ event: e, type: 'scrolltoupper' }))
        }  
      }

      if (typeof onScrollToLower === 'function') {
        if (scrollY) {
          const clientHeight = scrollRef.current.clientHeight;
          const scrollHeight = scrollRef.current.scrollHeight;
          const currentLower = scrollHeight - clientHeight - top;
          const prevLower = scrollHeight - clientHeight - topPreRef.current;
          if ((currentLowerThreshold > prevLower && currentLowerThreshold <= currentLower) || (currentLowerThreshold < prevLower && currentLowerThreshold >= currentLower)) {
            onScrollToLower(formatSyntheticEvent({ event: e, type: 'scrolltolower' }))
          }
        }
        if (scrollX) {
          const clientWidth = scrollRef.current.clientWidth;
          const scrollWidth = scrollRef.current.scrollWidth;
          const currentLower = scrollWidth - clientWidth - left;
          const prevLower = scrollWidth - clientWidth - leftPreRef.current;
          if ((currentLowerThreshold > prevLower && currentLowerThreshold <= currentLower) || (currentLowerThreshold < prevLower && currentLowerThreshold >= currentLower)) {
            onScrollToLower(formatSyntheticEvent({ event: e, type: 'scrolltolower' }))
          }
        }
      }

      topPreRef.current = top;
      leftPreRef.current = left;
    }
  }

  useLayoutEffect(() => {
    if (scrollY && scrollRef.current && scrollTop != null) {
      if (scrollWithAnimation) {
        scrollToWithAnimation(scrollRef.current, scrollTop, 300, true);
      } else {
        scrollRef.current.scrollTop = scrollTop;
      }
    }
  }, [scrollTop, scrollY, scrollWithAnimation])

  useLayoutEffect(() => {
    if (scrollX && scrollRef.current && scrollLeft != null) {
       if (scrollWithAnimation) {
        scrollToWithAnimation(scrollRef.current, scrollLeft, 300, false);
      } else {
        scrollRef.current.scrollLeft = scrollLeft;
      }
    }
  }, [scrollLeft, scrollX, scrollWithAnimation])

  useLayoutEffect(() => {
    if (scrollRef.current && scrollIntoView) {
      const $target = scrollRef.current.querySelector(`#${scrollIntoView}`)
      if ($target) {
        $target.scrollIntoView({ behavior: scrollWithAnimation ? 'smooth' : 'auto', block: 'start', inline: 'start' })
      }
    }
  }, [scrollIntoView, scrollWithAnimation])


  return (
    <div {...handleProps} className={handleClassName(styles.container)}>
      <div ref={scrollRef} className={styles.container} onScroll={handleScroll} style={{ overflow }}>
        <ScrollViewContext.Provider value={context}>
          {children}
        </ScrollViewContext.Provider>
      </div>
    </div>
  )
}

export default ScrollView;
