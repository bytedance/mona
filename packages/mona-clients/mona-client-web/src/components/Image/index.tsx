import React, { useContext, useEffect, useRef, useState } from 'react';
import { ImageProps } from '@bytedance/mona';
import cs from 'classnames';
import { ScrollViewContext } from '../ScrollView';
import styles from './index.module.less';
import { useHandlers } from '../hooks';
import { formatSyntheticEvent } from '../utils';

const ImageComponent: React.FC<ImageProps> = ({ children, src, mode = 'scaleToFill', lazyLoad = false, onError, onLoad, ...restProps }) => {
  const { handleClassName, ...handlerProps} = useHandlers(restProps);
  const [url, setUrl] = useState('');
  const [load, setLoad] = useState(!lazyLoad);
  const ref = useRef<HTMLDivElement | null>(null);
  const loadedRef = useRef(!lazyLoad);

  const context = useContext(ScrollViewContext);

  const handleLoad = (e: React.SyntheticEvent<HTMLImageElement, Event>) => {
    setUrl(src || '');
    if (typeof onLoad === 'function') {
      onLoad(formatSyntheticEvent({ event: e }));
    }
  }

  const handleError = (e: React.SyntheticEvent<HTMLImageElement, Event>) => {
    if (typeof onError === 'function') {
      onError(formatSyntheticEvent({ event: e }));
    }
  }

  // only run when init
  useEffect(() => {
    const callback = (scrollEle: Element) => {
      if (!loadedRef.current) {
        const { top: offsetTop, left: offsetLeft } = scrollEle.getBoundingClientRect();
        const clientHeight = scrollEle.clientHeight;
        const clientWidth = scrollEle.clientWidth;
        const rect = ref.current?.getBoundingClientRect();
        const top = rect?.top || 0;
        const left = rect?.left || 0;
        if ((offsetTop + clientHeight) >= top && (offsetLeft + clientWidth) >= left) {
          loadedRef.current = true;
          setLoad(true);
        }
      }
    }
    
    if (lazyLoad) {
      context?.registerScrollCallback(callback);
    }
    return () => {
      context?.unregisterScrollCallback(callback);
    }
  }, [lazyLoad])
  
  return (
    <div ref={ref} {...handlerProps} className={handleClassName(styles.wrapper)}>
      {load ? <img className={styles.hidden} src={src} onLoad={handleLoad} onError={handleError} /> : null}
      {url ? <div style={{ backgroundImage: `url("${url}")` }} className={cs(styles.image, styles[mode.replace(/\s/, '')])}></div> : null}
    </div>
  )
}

export default ImageComponent;
