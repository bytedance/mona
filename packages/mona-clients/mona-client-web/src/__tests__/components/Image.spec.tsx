import React from 'react';
import { mount } from 'enzyme';
import Image from '../../components/Image';
import mountTest from '../../../../../tests/shared/mountTest';
import handlerTest from '../../../../../tests/shared/handlerTest';
import { ImageProps } from '@bytedance/mona';
import ScrollView from '../../components/ScrollView';

describe('web component: Image', () => {
  mountTest(Image)
  handlerTest({ Component: Image });

  it('should render correctly', () => {
    expect(mount(<Image src="" />).render()).toMatchSnapshot();
  })
  
  it('should have right properties', () => {
    const props: ImageProps = {
      src: 'https://baidu.com',
      mode: 'bottom',
      lazyLoad: true,
      onError: () => {},
      onLoad: () => {},
      className: 'class-name',
      style: { color: 'red' },
    }
    const wrapper = mount(<Image {...props} />);
    expect(wrapper.render()).toMatchSnapshot();
  })

  // TODO: finish this
  it('should lazy load image when set lazyLoad: true', () => {
    const wrapper = mount(
      <ScrollView>
        <Image src='https://lf3-cm.ecombdstatic.com/obj/ecom-open-butler/prod/temp/90fab59c62dc11ecb1237cd30a502d1a.png' lazyLoad={true} />
      </ScrollView>
    );
    
    let counter = 0;
    jest.spyOn(Element.prototype, 'clientHeight', 'get').mockImplementation(() => 100);
    jest.spyOn(Element.prototype, 'clientWidth', 'get').mockImplementation(() => 100);
    // @ts-ignore
    jest.spyOn(Element.prototype, 'getBoundingClientRect').mockImplementation(() => {
      counter++;
      if (counter === 1) {
        return { top: 100, left: 0 }
      } else if (counter === 3) {
        return { top: 200, left: 0 }
      } else {
        return { top: 300, left: 0 }
      }
    })
    wrapper.find('div').at(1).simulate('scroll');
    expect(wrapper.find('img')).toHaveLength(0);
    wrapper.find('div').at(1).simulate('scroll');
    expect(wrapper.find('img')).toHaveLength(1);
  })

  // it('should trigger onError when image load failed', () => {
  //   const onLoad = jest.fn();
  //   const onError = jest.fn();

  //   const wrapper = mount(<Image onLoad={onLoad} src="aaa" />)
  //   expect(wrapper.find('img')).toHaveLength(1);
  //   wrapper.update();
  //   expect(onLoad).not.toHaveBeenCalled();
  //   expect(onError).toHaveBeenCalled();
  // })
})
