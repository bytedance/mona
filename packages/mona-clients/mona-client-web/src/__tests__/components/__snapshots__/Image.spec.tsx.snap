// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`web component: Image should have right properties 1`] = `
LoadedCheerio {
  "0": <div
    class="wrapper class-name"
    style="color: red;"
  />,
  "_root": LoadedCheerio {
    "0": Document {
      "children": Array [
        <html>
          <head />
          <body />
        </html>,
      ],
      "endIndex": null,
      "next": null,
      "parent": null,
      "prev": null,
      "startIndex": null,
      "type": "root",
      "x-mode": "quirks",
    },
    "_root": [Circular],
    "length": 1,
    "options": Object {
      "decodeEntities": true,
      "xml": false,
    },
  },
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "xml": false,
  },
}
`;

exports[`web component: Image should render correctly 1`] = `
LoadedCheerio {
  "0": <div
    class="wrapper"
  >
    <img
      class="hidden"
      src=""
    />
  </div>,
  "_root": LoadedCheerio {
    "0": Document {
      "children": Array [
        <html>
          <head />
          <body />
        </html>,
      ],
      "endIndex": null,
      "next": null,
      "parent": null,
      "prev": null,
      "startIndex": null,
      "type": "root",
      "x-mode": "quirks",
    },
    "_root": [Circular],
    "length": 1,
    "options": Object {
      "decodeEntities": true,
      "xml": false,
    },
  },
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "xml": false,
  },
}
`;
